dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -5962,6 +6000,640 @@ bool operator<<(Serializer& ser, const KeyOnly<DDS::DynamicData_ptr>& key)
+    CORBA::LongLong val = 0;
+    rc = value->get_int64_value(val, id);
+    if (!check_rc(rc, id, disc_tk, "vwrite_discriminator")) {
+      return;
+    }
+    disc_val = static_cast<CORBA::Long>(val);
+    vw.write_int64(val);
+    break;
+  }
+  case TK_UINT64: {
+    CORBA::ULongLong val = 0;
+    rc = value->get_uint64_value(val, id);
+    if (!check_rc(rc, id, disc_tk, "vwrite_discriminator")) {
+      return;
+    }
+    disc_val = static_cast<CORBA::Long>(val);
+    vw.write_uint64(val);
+    break;
+  }
+  default:
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_discriminator: Invalid discriminator type %C\n",
+                 typekind_to_string(disc_tk)));
+    }
+  }
+}
+
+void vwrite_union(ValueWriter& vw, DDS::DynamicData_ptr value, const DDS::DynamicType_var& dt)
+{
+  vw.begin_union();
+
+  // Discriminator
+  DDS::DynamicTypeMember_var dtm;
+  DDS::ReturnCode_t rc = dt->get_member(dtm, XTypes::DISCRIMINATOR_ID);
+  if (rc != DDS::RETCODE_OK) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_union: get_member failed (%C)\n",
+                 retcode_to_string(rc)));
+    }
+    return;
+  }
+  DDS::MemberDescriptor_var disc_md;
+  rc = dtm->get_descriptor(disc_md);
+  if (rc != DDS::RETCODE_OK) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_union: get_descriptor failed (%C)\n",
+                 retcode_to_string(rc)));
+    }
+    return;
+  }
+  vw.begin_discriminator();
+  CORBA::Long disc_val = 0;
+  vwrite_discriminator(vw, value, disc_md, disc_val);
+  vw.end_discriminator();
+
+  // Selected branch
+  bool has_branch = false;
+  DDS::MemberDescriptor_var selected_md;
+  rc = XTypes::get_selected_union_branch(dt, disc_val, has_branch, selected_md);
+  if (rc != DDS::RETCODE_OK) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_union: get_selected_union_branch failed (%C)\n",
+                 retcode_to_string(rc)));
+    }
+    return;
+  }
+  if (has_branch) {
+    vw.begin_union_member(selected_md->name());
+    vwrite_member(vw, value, selected_md);
+    vw.end_union_member();
+  }
+
+  vw.end_union();
+}
+
+void vwrite_element(ValueWriter& vw, DDS::DynamicData_ptr value,
+                    const DDS::DynamicType_var& elem_dt, CORBA::ULong idx)
+{
+  const DDS::MemberId id = value->get_member_id_at_index(idx);
+  if (id == XTypes::MEMBER_ID_INVALID) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_element: get_member_id_at_index %u failed\n", idx));
+    }
+    return;
+  }
+  vwrite_item(vw, value, id, elem_dt);
+}
+
+// When we're sure that the concrete DynamicData class has implemented get_*_values operations
+// for primitive sequences, we can use ValueWriter's write_*_array functions.
+// For now, vwrite for array and sequence will write elements one by one.
```

Comment:
Which implementations are missing those functions?  One strategy here would be to call the _values function and if it returns an error fall back to the one-at-a-time method.

Reply:
`DynamicDataImpl` and `DynamicDataAdapter` haven't had these methods yet. Yeah, the fall back strategy sounds good to me.

Reply:
Looking at this comment again, it was more of an optimization for primitive sequence/array with a XcdrValueWriter when we support it because `Serializer` has functions to write arrays and they're more efficient. For `JsonValueWriter` and `PrinterValueWriter`, their write_*_array functions also write elements one at a time. So I'd prefer to defer this to when we implement a XcdrValueWriter, preferably in #4392.